File: 56. Merge Intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    vector<Interval> merge(vector<Interval>& intervals) {
        vector<Interval> res;
        vector<Interval>& a= intervals;
        int n = a.size();
        
        sort(a.begin(), a.end(), [&](const Interval& a, const Interval& b){
            return a.start < b.start;    
        });
        
        for (int i = 0; i < n - 1;)
        {
            if (a[i].end >= a[i+1].start)
            {
                a[i].start = min(a[i].start, a[i+1].start);
                a[i].end = max(a[i].end, a[i+1].end);
                a.erase(a.begin()+i+1);
                n--;
            }
            else
                i++;
        }
        return a;
    }
};

File: go/src/leetcode/56. Merge Intervals/56. Merge Intervals.go
package main

import (
	"fmt"
	"sort"
)

func merge(intervals [][]int) [][]int {
	sort.Slice(intervals, func(i, j int) bool {
		return intervals[i][0] <= intervals[j][0]
	})
	erase := make(map[int]int)
	n := len(intervals)
	last := intervals[0]
	for i := 1; i < n; i++ {
		if last[1] >= intervals[i][0] {
			last[1] = max(last[1], intervals[i][1])
			erase[i]++
		} else {
			last = intervals[i]
		}
	}

	ans := [][]int{}
	for i := 0; i < n; i++ {
		if erase[i] != 0 {
			continue
		}
		ans = append(ans, intervals[i])
	}

	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(merge([][]int{{1,3},{2,6},{8,10},{15,18}}))
	fmt.Println(merge([][]int{{1,4},{4,5}}))
}
